### PR TITLE
Exposing Input Key Binding To JavaScript

### PR DESCRIPTION
**Pull request by SamGondelman**
Enables users to remap key bindings and adjust action sensitivities via javascript

More functionality will be added as more devices are moved to this new system and desired functionality becomes evident, but right now you should be able to do pretty much anything with the keyboard and mouse bindings (see the provided example script).

Known Issue:  Adjusting sensitivity for directional motion won't have any effect because the acceleration curve is hardcoded elsewhere, but that's unrelated to actually exposing the functions.  You can test sensitivity adjustment by changing something like the "a" and "d" keys

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: dd12e4740e036895929ddd42e63eab9caab6fab0 available at: https://gist.github.com/99235bd1dbb500dca9e2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5085.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5085.dmg

File: interface/src/devices/KeyboardMouseDevice.cpp [hunk 1]
```diff
@@ -159,9 +159,25 @@ void KeyboardMouseDevice::registerToUserInputMapper(UserInputMapper& mapper) {
     // Grab the current free device ID
     _deviceID = mapper.getFreeDeviceID();
 
-    auto proxy = UserInputMapper::DeviceProxy::Pointer(new UserInputMapper::DeviceProxy());
+    auto proxy = UserInputMapper::DeviceProxy::Pointer(new UserInputMapper::DeviceProxy("Keyboard"));
     proxy->getButton = [this] (const UserInputMapper::Input& input, int timestamp) -> bool { return this->getButton(input._channel); };
     proxy->getAxis = [this] (const UserInputMapper::Input& input, int timestamp) -> float { return this->getAxis(input._channel); };
+    proxy->getAvailabeInputs = [this] () -> QVector<UserInputMapper::InputPair> {
+        QVector<UserInputMapper::InputPair> availableInputs;
+        for (int i = 0; i <= 9; i++) {
```
Comment (ZappoMan) [inline interface/src/devices/KeyboardMouseDevice.cpp]: coding standard: no magic numbers

File: interface/src/devices/KeyboardMouseDevice.cpp [hunk 2]
```diff
@@ -159,9 +159,25 @@ void KeyboardMouseDevice::registerToUserInputMapper(UserInputMapper& mapper) {
...
+            availableInputs.append(UserInputMapper::InputPair(makeInput(Qt::Key(48 + i)), QKeySequence(Qt::Key(48 + i)).toString()));
```
Comment (ZappoMan) [inline interface/src/devices/KeyboardMouseDevice.cpp]: coding standard: no magic numbers

File: interface/src/devices/KeyboardMouseDevice.cpp [hunk 3]
```diff
@@ -159,9 +159,25 @@ void KeyboardMouseDevice::registerToUserInputMapper(UserInputMapper& mapper) {
...
+        }
+        for (int i = 0; i < 26; i++) {
```
Comment (ZappoMan) [inline interface/src/devices/KeyboardMouseDevice.cpp]: coding standard: no magic numbers

File: interface/src/devices/KeyboardMouseDevice.cpp [hunk 4]
```diff
@@ -159,9 +159,25 @@ void KeyboardMouseDevice::registerToUserInputMapper(UserInputMapper& mapper) {
...
+            availableInputs.append(UserInputMapper::InputPair(makeInput(Qt::Key(65 + i)), QKeySequence(Qt::Key(65 + i)).toString()));
```
Comment (ZappoMan) [inline interface/src/devices/KeyboardMouseDevice.cpp]: coding standard: no magic numbers

File: interface/src/scripting/ControllerScriptingInterface.cpp [hunk 5]
```diff
@@ -28,6 +28,90 @@ ControllerScriptingInterface::ControllerScriptingInterface() :
...
+    QScriptValue obj = engine->newObject();
+    obj.setProperty("device", input.getDevice());
+    obj.setProperty("channel", input.getChannel());
+    obj.setProperty("type", input._type);
+    obj.setProperty("id", input.getID());
+    return obj;
+}
+
+void inputFromScriptValue(const QScriptValue& object, UserInputMapper::Input& input) {
+    input._device = object.property("device").toUInt16();
```
Comment (ZappoMan) [inline interface/src/scripting/ControllerScriptingInterface.cpp]: something seems wrong about this... the "_" is supposed to generally be used for private members... and we should expose them through getter/setters... I suspect you didn't write the UserInputMapper::Input and so this was probably how you found it... but if there are getter/setters available to use those would be preferred.

File: interface/src/ui/UserInputMapper.h [hunk 6]
```diff
@@ -123,8 +135,14 @@ class UserInputMapper : public QObject {
...
         NUM_ACTIONS,
     };
+    
+    std::vector<QString> _actionNames = std::vector<QString>(NUM_ACTIONS);
+    void createActionNames();
 
+    QVector<Action> getAllActions();
+    QString getActionName(Action action) { return UserInputMapper::_actionNames[(int) action]; }
     float getActionState(Action action) const { return _actionStates[action]; }
+//    QVector<InputChannel>
```
Comment (ZappoMan) [inline interface/src/ui/UserInputMapper.h]: dead code?

File: examples/example/scripts/controllerScriptingExamples.js [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+//
+//  Created by Sam Gondelman on 6/2/15
+//  Copyright 2015 High Fidelity, Inc.
+//
+//  Distributed under the Apache License, Version 2.0.
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+
+// Resets every device to its default key bindings:
+Controller.resetAllDeviceBindings()
```
Comment (ctrlaltdavid) [inline examples/example/scripts/controllerScriptingExamples.js]: Missing ";"s

**Comment by hifi-gustavo** [comment]
Build console output for commit: b1b2d1f85cf54f9d0bf65a6674bb4c4c3fdc0fb6 available at: https://gist.github.com/99235bd1dbb500dca9e2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5085.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5085.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d3a4eec5c5a935bfacd56221672829f246c18e2f available at: https://gist.github.com/99235bd1dbb500dca9e2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5085.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5085.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0769593ca89c74a0b84f38e8066e2b43cfa5e42f available at: https://gist.github.com/99235bd1dbb500dca9e2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5085.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5085.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0e8ec81837f56f4a211fc843c3a51c4aa168308d available at: https://gist.github.com/99235bd1dbb500dca9e2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5085.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5085.dmg

**Pull request merged**